### ISSUE TITLE
F# records in reference assembly (/ref folder) are missing properties except for the last record in file [regression with 7.0.100-rc.2.22477.23 SDK ]

### ISSUE THREAD

**Issue by buvinghausen** [dotnet/fsharp#14088]
After downloading and installing the latest 7.0.100-rc.2.22477.23 SDK I can no longer compile my C# libraries because if any code appears after the F# record declaration the compiler will throw a CS0117 error if you try to set a property on a CLIMutable record and/or a CS1061 error if you try to get any property on a record from C#.

**Repro steps**

Provide the steps required to reproduce the problem:

1. Create an F# class library with 2 record types
2. Create a C# console app that references the F# library
3. Initialize the first record declaration and attempt to access the properties in the C# console

[Clone here to speed up the repro](https://github.com/buvinghausen/Net7FSharpSnafu)

**Expected behavior**

Expect to be able to still use F# records in C# after F# code appears after them

**Actual behavior**

CS0117 if you attempt to set/initialize an F# record property where F# code appears after the record declaration
CS1061 if you attempt to get the value from an F# record property where F# code appears after the record declaration

**Known workarounds**

```xml
<ProduceReferenceAssembly>false</ProduceReferenceAssembly>
```

**Related information**

7.0.100-rc.2.22477.23 SDK on both Windows & Ubuntu

**Comment by T-Gro** [comment]
The properties (in the C# view of things) are missing alltogether:

![image](https://user-images.githubusercontent.com/46543583/195292826-17406498-be09-49df-aaf3-14a73f91f58d.png)

**Comment by vzarytovskii** [comment]
One thing rc2 has on by default is producing refassemblies for net7 projects, this might be it.

We might want to roll it back until fully tested.

**Comment by T-Gro** [comment]
Indeed, the .dll within /ref folder is missing the properties for all but the last record:

![image](https://user-images.githubusercontent.com/46543583/195299675-fe4c2833-fb68-4c8e-9384-7bf03b8b89f7.png)

**Comment by T-Gro** [comment]
# Workaround

To overcome this, please disable generation of F# reference assemblies (is on by default). That is, put the following in the .fsproj.
Note that the files are still cached and ref from the obj/Debug/ref folder, so I had to close VS, delete bin and obj, and then reopen.

After that, the project builds and runs.

```
  <PropertyGroup>
    <TargetFramework>net7.0</TargetFramework>
    <GenerateDocumentationFile>true</GenerateDocumentationFile>
    <ProduceReferenceAssembly>false</ProduceReferenceAssembly>
  </PropertyGroup>
```

**Comment by vzarytovskii** [comment]
@T-Gro can you take a look at it please?

Ping me on teams if you'll need need help with figuring out how refassembly generation works.

**Comment by vzarytovskii** [comment]
`canGenPropertyDef` in `ilwrite.fs` is the place where we decide whether we want to emit property or not.

**Comment by T-Gro** [comment]
The issue is curious at the IL gen level:

It only occurs when I have records with equaly-named properties.
What happens is that the properties in generated IL code get assigned to the last class in file, from all records across the module.

Even the screenshot above makes this clear - I had three records, and it generated each prop three times for the very last record.

**Comment by vzarytovskii** [comment]
> The issue is curious at the IL gen level:
> 
> It only occurs when I have records with equaly-named properties.
> What happens is that the properties in generated IL code get assigned to the last class in file, from all records across the module.
> 
> Even the screenshot above makes this clear - I had three records, and it generated each prop three times for the very last record.

Huh, that's interesting, wondering if it's because we don't take types into account when we decide how to generate them?

**Comment by vzarytovskii** [comment]
@buvinghausen while we're investigating, setting ProduceReferenceAssembly msbuild property can be used as a workaround. Sorry for the inconvenience.

**Comment by vzarytovskii** [comment]
The fix should be in the next released of 17.4 and .net7

**Comment by buvinghausen** [comment]
Thanks @vzarytovskii glad it will make it into the .NET 7.0 RTM cycle!

**Comment by vzarytovskii** [comment]
This was fixed and inserted to 6.0 and 7.0
